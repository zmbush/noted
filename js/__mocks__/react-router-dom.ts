// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
// eslint-disable-next-line import/no-import-module-exports
import type { NavigateFunction, To, NavigateOptions } from 'react-router-dom';

const ReactRouterDom = jest.requireActual('react-router-dom');

const useNavigate = (): NavigateFunction => {
  try {
    return ReactRouterDom.useNavigate();
  } catch {
    return (_to: To | number, _options?: NavigateOptions) => {};
  }
};

module.exports = { ...ReactRouterDom, useNavigate };
