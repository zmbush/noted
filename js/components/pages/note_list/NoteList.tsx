// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import Fuse from 'fuse.js';
import memoize from 'memoize-one';

import * as React from 'react';
import { useSelector } from 'react-redux';

import { Add as AddIcon } from '@mui/icons-material';
import { Button, Grid, styled } from '@mui/material';

import Note from 'components/note/Note';
import {
  getFilteredSearchIndex,
  getNoteEntities,
  getSortedNoteIds,
  getSubNotes,
} from 'data/notes/selectors';
import { AppState } from 'data/store';
import { NoteWithTags } from 'data/types';

type Props = {
  search: string;
  depth: number;
  createFromSearch?: (e: React.SyntheticEvent) => void;
  renderOnly?: Set<number>;
  noteViewFilter: { [id: number]: boolean } | null;
  parent_note_id: number | null;
  width?: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
};

const GridItem = styled(Grid)({
  paddingTop: 0,
  '@media print': {
    padding: '0 !important',
  },
});

const getFuse = memoize(
  (index: { [id: string]: NoteWithTags }) =>
    new Fuse(Object.values(index), {
      distance: 100,
      keys: [
        {
          name: 'title',
          weight: 1.0,
        },
        {
          name: 'tags',
          weight: 1.0,
        },
        {
          name: 'body',
          weight: 0.5,
        },
      ],
      location: 0,
      shouldSort: true,
      threshold: 0.4,
    }),
);

const NoteList = ({
  search,
  depth,
  createFromSearch,
  renderOnly,
  width = 12,
  noteViewFilter,
  parent_note_id: parentNoteId,
}: Props) => {
  const notesIn = useSelector<AppState, { [id: string]: NoteWithTags }>((state) =>
    renderOnly ? getNoteEntities(state) : getSubNotes(state, { noteId: parentNoteId }),
  );
  const searchIndex = useSelector((state) =>
    getFilteredSearchIndex(state, { noteId: parentNoteId }),
  );
  const sortedIds = useSelector(getSortedNoteIds);

  let notes = notesIn;
  if (renderOnly) {
    notes = Object.fromEntries(
      Object.entries(notes).filter(([id, _note]) => renderOnly.has(parseInt(id, 10))),
    );
  }

  if (search !== '') {
    const elements = [];
    const results = getFuse(searchIndex).search(search);

    if (depth === 1 && (results.length === 0 || notes[results[0].item.id].title !== search)) {
      elements.push(
        <GridItem item key='new' xs={width}>
          <Button
            variant='contained'
            color='primary'
            sx={{
              margin: 1,
              '@media print': {
                display: 'none',
              },
            }}
            onClick={createFromSearch}
          >
            <AddIcon
              sx={{
                marginRight: 1,
                fontSize: 20,
              }}
            />
            Add {search}
          </Button>
        </GridItem>,
      );
    }

    results.forEach((result) => {
      const { id } = result.item;
      if (!(id in notes)) {
        // eslint-disable-next-line no-console
        console.log('Note ', id, ' not found');
      } else {
        elements.push(
          <GridItem item key={id} xs={width}>
            <Note note={notes[id]}>
              <NoteList
                parent_note_id={id}
                noteViewFilter={noteViewFilter}
                depth={depth + 1}
                search={search}
              />
            </Note>
          </GridItem>,
        );
      }
    });
    // eslint-disable-next-line react/jsx-no-useless-fragment
    return <>{elements}</>;
  }

  return (
    <>
      {sortedIds.map((id) => {
        if (id in notes) {
          const n = notes[id];
          if (!noteViewFilter || noteViewFilter[id]) {
            return (
              <GridItem item key={n.id} xs={width}>
                <Note note={n}>
                  <NoteList
                    parent_note_id={n.id}
                    noteViewFilter={noteViewFilter}
                    depth={depth + 1}
                    search={search}
                  />
                </Note>
              </GridItem>
            );
          }
        }
        return null;
      })}
    </>
  );
};

export default NoteList;
