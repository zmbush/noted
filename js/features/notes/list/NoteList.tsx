// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import * as React from 'react';
import { useSelector } from 'react-redux';

import { Grid, styled } from '@mui/material';

import { AppState } from 'features/redux/store';

import { NoteWithTags } from 'data/types';

import Note from '../view/Note';

import { getNoteEntities, getSortedNoteIds, getSubNotes } from './selectors';

type Props = {
  depth: number;
  renderOnly?: Set<number>;
  noteViewFilter: { [id: number]: boolean } | null;
  parent_note_id: number | null;
  width?: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
};

const GridItem = styled(Grid)({
  paddingTop: 0,
  '@media print': {
    padding: '0 !important',
  },
});

const NoteList = ({
  depth,
  renderOnly,
  width = 12,
  noteViewFilter,
  parent_note_id: parentNoteId,
}: Props) => {
  const notesIn = useSelector<AppState, { [id: string]: NoteWithTags }>((state) =>
    renderOnly ? getNoteEntities(state) : getSubNotes(state, { noteId: parentNoteId }),
  );
  const sortedIds = useSelector(getSortedNoteIds);

  let notes = notesIn;
  if (renderOnly) {
    notes = Object.fromEntries(
      Object.entries(notes).filter(([id, _note]) => renderOnly.has(parseInt(id, 10))),
    );
  }

  return (
    <>
      {sortedIds.map((id) => {
        if (id in notes) {
          const n = notes[id];
          if (!noteViewFilter || noteViewFilter[id]) {
            return (
              <GridItem item key={n.id} xs={width}>
                <Note note={n}>
                  <NoteList
                    parent_note_id={n.id}
                    noteViewFilter={noteViewFilter}
                    depth={depth + 1}
                  />
                </Note>
              </GridItem>
            );
          }
        }
        return null;
      })}
    </>
  );
};

export default NoteList;
