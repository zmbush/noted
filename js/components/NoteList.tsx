// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import * as React from 'react';
import Fuse from 'fuse.js';
import AddIcon from '@material-ui/icons/Add';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import Note, { InnerNote } from 'components/Note';
import classNames from 'classnames';
import { NoteData, AppState } from 'data/types';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { withStyles, createStyles, WithStyles } from '@material-ui/core/styles';
import {
  LinkIdMap,
  getFilteredSearchIndex,
  getSortedNoteIds,
} from 'data/selectors';
import { connect } from 'react-redux';

const styles = (theme: Theme) =>
  createStyles({
    newButton: {
      margin: theme.spacing.unit,
      '@media print': {
        display: 'none',
      },
    },
    item: {
      '@media print': {
        padding: '0 !important',
      },
    },
    leftIcon: {
      marginRight: theme.spacing.unit,
    },
    iconSmall: {
      fontSize: 20,
    },
  });

interface Props extends WithStyles<typeof styles>, RouteComponentProps {
  notes: Map<number, NoteData>;
  searchIndex: Map<number, NoteData>;
  sortedIds: number[];
  search: string;
  depth: number;
  updateNote: (note?: NoteData) => void;
  deleteNote: (id: number) => void;
  createFromSearch?: (e: React.SyntheticEvent) => void;
  firstNoteRef?: React.RefObject<InnerNote>;
  renderOnly?: Set<number>;
  width?:
    | false
    | 'auto'
    | true
    | 1
    | 2
    | 3
    | 4
    | 5
    | 6
    | 7
    | 8
    | 9
    | 10
    | 11
    | 12;
}

class NoteList extends React.Component<Props> {
  static defaultProps = {
    width: 12 as 12,
  };

  render() {
    const { classes } = this.props;
    let { notes } = this.props;

    if (this.props.renderOnly) {
      notes = new Map(
        Array.from(notes.entries()).filter(([id, note]) =>
          this.props.renderOnly.has(id)
        )
      );
    }

    if (this.props.search != '') {
      let fuse = new Fuse(Array.from(this.props.searchIndex.values()), {
        distance: 100,
        keys: [
          {
            name: 'title',
            weight: 1.0,
          },
          {
            name: 'tags',
            weight: 1.0,
          },
          {
            name: 'body',
            weight: 0.5,
          },
        ],
        id: 'id',
        location: 0,
        matchAllTokens: true,
        maxPatternLength: 32,
        minMatchCharLength: 1,
        shouldSort: true,
        threshold: 0.4,
        tokenize: true,
      });

      let elements = [];
      let results = (fuse.search(this.props.search) as any) as string[];

      if (
        this.props.depth == 1 &&
        (results.length == 0 ||
          notes.get(parseInt(results[0], 10)).title != this.props.search)
      ) {
        elements.push(
          <Grid item key='new' className={classes.item} xs={this.props.width}>
            <Button
              variant='contained'
              color='primary'
              className={classes.newButton}
              onClick={this.props.createFromSearch}
            >
              <AddIcon
                className={classNames(classes.leftIcon, classes.iconSmall)}
              />
              Add {this.props.search}
            </Button>
          </Grid>
        );
      }

      let i = 0;
      for (let idStr of results) {
        let id = parseInt(idStr, 10);
        if (!notes.has(id)) {
          console.log('Note ', id, ' not found');
        } else {
          elements.push(
            <Grid item key={id} className={classes.item} xs={this.props.width}>
              <Note
                depth={this.props.depth + 1}
                note={notes.get(id)}
                search={this.props.search}
                updateNote={this.props.updateNote}
                deleteNote={this.props.deleteNote}
                innerRef={i == 0 ? this.props.firstNoteRef : null}
              />
            </Grid>
          );
          i++;
        }
      }
      return elements;
    } else {
      let result = [];
      for (let id of this.props.sortedIds) {
        if (notes.has(id)) {
          const n = notes.get(id);
          result.push(
            <Grid
              item
              key={n.id}
              className={classes.item}
              xs={this.props.width}
            >
              <Note
                depth={this.props.depth + 1}
                note={n}
                updateNote={this.props.updateNote}
                deleteNote={this.props.deleteNote}
                search={this.props.search}
              />
            </Grid>
          );
        }
      }
      return result;
    }
  }
}

const mapStateToProps = (
  state: AppState,
  props: { parent_note_id: number }
) => ({
  searchIndex: getFilteredSearchIndex(state, { note_id: props.parent_note_id }),
  sortedIds: getSortedNoteIds(state),
});

export default withRouter(
  connect(mapStateToProps)(withStyles(styles)(NoteList))
);
