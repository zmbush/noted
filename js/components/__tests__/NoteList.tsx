// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { shallow } from 'enzyme';

import * as React from 'react';
import * as ReactRedux from 'react-redux';
import * as ReactRouterDomOriginal from 'react-router-dom';

import { testState } from 'data/utils.forTesting';

import NoteList from '../NoteList';

jest.mock('react-router');
const ReactRouterDom = ReactRouterDomOriginal as jest.Mocked<typeof ReactRouterDomOriginal>;

ReactRouterDom.useMatch.mockReturnValue(null);
// eslint-disable-next-line no-import-assign
jest.spyOn(ReactRedux, 'useSelector').mockImplementation((r) => r(testState));

describe('<NoteList />', () => {
  test('matches snapshot', () => {
    const node = shallow(
      <NoteList parent_note_id={0} depth={1} notes={testState.notes} search='' />,
    );

    expect(node).toMatchSnapshot();
  });

  test('matches second snapshot', () => {
    const node = shallow(
      <NoteList parent_note_id={0} depth={1} notes={testState.notes} search='SingleNote 2' />,
    );
    expect(node).toMatchSnapshot();
  });
});
