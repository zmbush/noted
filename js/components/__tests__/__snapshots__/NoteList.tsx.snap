// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteList /> matches second snapshot 1`] = `
<Fragment>
  <Styled(ForwardRef(Grid))
    item={true}
    key="new"
    xs={12}
  >
    <ForwardRef(Button)
      color="primary"
      sx={
        Object {
          "@media print": Object {
            "display": "none",
          },
          "margin": 1,
        }
      }
      variant="contained"
    >
      <Memo(ForwardRef(AddIcon))
        sx={
          Object {
            "fontSize": 20,
            "marginRight": 1,
          }
        }
      />
      Add 
      SingleNote 2
    </ForwardRef(Button)>
  </Styled(ForwardRef(Grid))>
</Fragment>
`;

exports[`<NoteList /> matches snapshot 1`] = `
<Fragment>
  <Styled(ForwardRef(Grid))
    item={true}
    key="1"
    xs={12}
  >
    <Note
      depth={2}
      note={
        Object {
          "archived": false,
          "body": "The Body",
          "created_at": "",
          "id": 1,
          "parent_note_id": 0,
          "pinned": false,
          "tags": Array [],
          "title": "Note 1",
          "updated_at": "",
          "user_id": -1,
        }
      }
      search=""
    />
  </Styled(ForwardRef(Grid))>
  <Styled(ForwardRef(Grid))
    item={true}
    key="4"
    xs={12}
  >
    <Note
      depth={2}
      note={
        Object {
          "archived": false,
          "body": "The Body",
          "created_at": "",
          "id": 4,
          "parent_note_id": 0,
          "pinned": false,
          "tags": Array [],
          "title": "Note 4",
          "updated_at": "",
          "user_id": -1,
        }
      }
      search=""
    />
  </Styled(ForwardRef(Grid))>
  <Styled(ForwardRef(Grid))
    item={true}
    key="2"
    xs={12}
  >
    <Note
      depth={2}
      note={
        Object {
          "archived": false,
          "body": "The Body",
          "created_at": "",
          "id": 2,
          "parent_note_id": 0,
          "pinned": false,
          "tags": Array [],
          "title": "Note 2",
          "updated_at": "",
          "user_id": -1,
        }
      }
      search=""
    />
  </Styled(ForwardRef(Grid))>
</Fragment>
`;
